kernel/kernel-impl/src/main/java/org/sakaiproject/authz/impl/SakaiSecurity.java

```diff
@@ -705,6 +725,74 @@ protected boolean checkAuthzGroups(String userId, String function, String entity
 			azgs = ref.getAuthzGroups(userId);
 		}
 
+		// need to know whether role swap is in effect, since we can't share the cache entry between sessions
+		// that are swapped and not swapped
+
+		String roleswap = null; // define the roleswap variable
+
+		// this version seems to be enough
+		if (azgs != null && userId != null && userId.equals(sessionManager().getCurrentSessionUserId())) {
```

Comment:
That's what the code currently does. One thing I see that should probably be changed is that I check for /site and /content, but anything else is still checked with getUserEffectiveRole. The code in DbAuthzGroupService does that, but it's pretty silly, since nothing else could match.